### PR TITLE
ANN bench options to specify CAGRA graph and dataset locations

### PR DESCRIPTION
**Pull request by cjnolet**
CAGRA index can be constructed using existing device_mdarrays, in which case just reference to these arrays are stored. 
This way allocations are managed outside the index, and we can customize how these allocations are made.

This PR
- modifies the CAGRA ANN bench wrapper to manage the allocations locally,
- add options for the json file to specify whether the graph / dataset is allocated in device / host_pinned / host_huge_page memory.

### CODE REVIEW

**Comment by tfeher** [comment]
Putting here Corey's original comments in the PR description:

>These are changes that were spearheaded by @tfeher for profiling and we think they've demonstrated enough performance benefit that we should incorporate them into RAFT. The big question is, how do we do this? What should the ideal design look like? This is the first time enabling a feature like this where only certain allocations are using pinned memory while other allocations aren't so we want to focus on a scalable and maintainable solution. 

After the latest changes, the PR does not change CAGRA API. It enables us to benchmark with data in different memory locations, and also demonstrates that it is possible to do so with the existing API.

File: cpp/bench/ann/src/raft/raft_cagra_wrapper.h
Comment (tfeher) [reply cpp/bench/ann/src/raft/raft_cagra_wrapper.h]: Debug comment, removed.

File: cpp/bench/ann/src/raft/raft_cagra_wrapper.h
Comment (tfeher) [reply cpp/bench/ann/src/raft/raft_cagra_wrapper.h]: Fixed.

File: cpp/bench/ann/src/common/cuda_huge_page_resource.hpp
Comment (tfeher) [reply cpp/bench/ann/src/common/cuda_huge_page_resource.hpp:34]: Updated the documentation.
> Is there a reason why it is not derived from rmm::mr::host_memory_resource?

The goal is to use this allocator in a [device_container_policy](https://github.com/rapidsai/raft/blob/branch-23.12/cpp/include/raft/core/device_container_policy.hpp#L177), and that requires a `device_memory_resource`.
What we have here is not ideal, more like a quick solution to avoid experimenting / benchmarking with such settings. A related issue https://github.com/rapidsai/raft/issues/1819.

**Review by tfeher — COMMENTED** [review]
Thanks @lowener for the review, I have addressed the issues.

**Review by lowener — APPROVED** [review]

File: cpp/bench/ann/src/raft/raft_cagra_wrapper.h
Comment (cjnolet) [inline cpp/bench/ann/src/raft/raft_cagra_wrapper.h:205]: Can you make an issue for this and reference it here so we can keep track of it?

File: cpp/bench/ann/src/common/cuda_huge_page_resource.hpp
Comment (cjnolet) [inline cpp/bench/ann/src/common/cuda_huge_page_resource.hpp:2]: I should have noticed this the first time around- dates for new files should be 2023.

File: cpp/bench/ann/src/common/cuda_pinned_resource.hpp
Comment (cjnolet) [inline cpp/bench/ann/src/common/cuda_pinned_resource.hpp:2]: Here too- 2023

**Review by cjnolet — COMMENTED** [review]
2 minor things. I can't approve my own PR but please feel free to approve it. Tagging @divyegala for approval as well.

File: cpp/bench/ann/src/raft/raft_benchmark.cu
Comment (cjnolet) [inline cpp/bench/ann/src/raft/raft_benchmark.cu:220]: Can you also update the Ann bench param tuning guide to add the men types?

**Review by divyegala — APPROVED** [review]

File: cpp/bench/ann/src/common/cuda_huge_page_resource.hpp
Comment (tfeher) [reply cpp/bench/ann/src/common/cuda_huge_page_resource.hpp:2]: fixed

File: cpp/bench/ann/src/raft/raft_benchmark.cu
Comment (tfeher) [reply cpp/bench/ann/src/raft/raft_benchmark.cu:220]: updated

**Review by tfeher — COMMENTED** [review]
Thanks @cjnolet for the review, I have addressed the issues.

File: cpp/bench/ann/src/common/cuda_huge_page_resource.hpp
Comment (tfeher) [inline cpp/bench/ann/src/common/cuda_huge_page_resource.hpp:75]: There is a request provide an alternative methods for allocation (without `mmap`) so that we can use Transparent Huge Pages. I will submit a follow up PR with these changes. I am thinking about the following design:

```c++
enum class AllocatorType { HostPinned, HostMmap, HostMalloc }

class host_device_allocator final : public rmm::mr::device_memory_resource {
  host_device_allocator(AllocatorType type): type_(type) {}
  //...
  void* do_alocate(std::size_t bytes, rmm::cuda_stream_view) override {
    switch(type_) {
      case HostPinned:
        cudaMallocHost(&ptr, bytes));
        return ptr;
      case HostMmap:
        ptr = mmap(...);
        //...
        return ptr;
      case HostMalloc:
        //...
    } 
  }
 private:
  AllocatorType type_;
//...
}
```
In this case we would have a single custom `mr` class that we would configure according to the benchmark input parameter. Let me know what do you think.

Comment (cjnolet) [reply cpp/bench/ann/src/common/cuda_huge_page_resource.hpp:75]: Hmm, I'll have to think about this. I don't think I generally mind the approach of having a `composite_memory_resource` or `delegating_memory_resource` but I think we should continue using RMM's abstraction layers everywhere possible instead of, for example, calling `cudaMallocHost` directly. My concern is a pretty easy fix, though, by just storing the corresponding memory_resource instances inside the `delegating_memory_resource` and delegating as necessary. That would maintain the compatibility w/ RMM (this could, for example, still be set as the `current_device_resource` if we so wanted).

Comment (cjnolet) [reply cpp/bench/ann/src/common/cuda_huge_page_resource.hpp:75]: Also cc'ing @harrism for his thoughts.

File: cpp/bench/ann/src/common/cuda_pinned_resource.hpp
Comment (cjnolet) [inline cpp/bench/ann/src/common/cuda_pinned_resource.hpp:39]: It would be nice to see these new memory resources worked back into RMM eventually.

File: cpp/bench/ann/src/raft/raft_benchmark.cu
Comment (cjnolet) [inline cpp/bench/ann/src/raft/raft_benchmark.cu:254]: Can we spell out `*_memory_type` to match the other properties that we parse for other algorithms? Will get confusing to set these if we use different spellings.

File: cpp/bench/ann/src/common/cuda_pinned_resource.hpp
Comment (tfeher) [reply cpp/bench/ann/src/common/cuda_pinned_resource.hpp:39]: RMM has a pinned memory resource. But it is a subclass of `host_memory_resource`. The problem is on our side: our API for [managing allocators for mdspan ](https://github.com/rapidsai/raft/blob/061c0cf7a50bff8600ca9ecef86edee81a354233/cpp/include/raft/core/device_container_policy.hpp#L177) requires a device_memory_resource. We would need to enable host allocator there, and then we could use `rmm::mr::host_memory_resource`.

File: cpp/bench/ann/src/common/cuda_huge_page_resource.hpp
Comment (tfeher) [reply cpp/bench/ann/src/common/cuda_huge_page_resource.hpp:75]: The class above is a subclass of `rmm::mr::device_memory_resource`. We are using it to allocate `mdarray`s that are accessible from both host and device. It is a workaround to unblock some tests.

I feel that RAFT's distinction between `host_mdarry` and `device_mdarray`  stands in the way. What our algorithm needs is an `mdarray` which we can configured with any allocator (host, device). In a few places our algorithm does not care whether the input is a `host_mdarray` or a `device_mdarray`, just passes the pointer to `cudaMemcpy`, to fetch a chunk of the array we are working on. 

How we allocate `mdarray` is described by a `ContainerPolicy` class. So probably the only thing we need is to implement a contaner policy that could be initialized by both `device_memory_resource` and `host_memory_resource`. Afterwards we can simply utilize existing RMM memory resources.

The only thing we are missing from RMM side is a memory resource that uses `mmap` and `madvise`.

**Comment by raydouglass** [comment]
Removing `ops-codeowners` from the required reviews since it doesn't seem there are any file changes that we're responsible for. Feel free to add us back if necessary.

File: cpp/bench/ann/src/common/cuda_pinned_resource.hpp
Comment (harrism) [reply cpp/bench/ann/src/common/cuda_pinned_resource.hpp:39]: We have a plan too add a device-accessible pinned memory resource. Ideally using the new cuda::memory_resource refactoring.

File: cpp/bench/ann/src/common/cuda_huge_page_resource.hpp
Comment (harrism) [reply cpp/bench/ann/src/common/cuda_huge_page_resource.hpp:75]: Look, it's fine to provide your own base-level `memory_resource` classes that derive from RMM resource base classes. The problem today is that there is no common base for host and device MRs. We are starting to refactor on top of `cuda::memory_resource` which abstracts memory `kind` as a template parameter, and adds `resource_ref` and `async_resource_ref` which have template parameters to specify memory kind. This means you can have a function that accepts any memory resource that is device accessible, host accessible, or both, or whatever kind you can define. And we can start to reuse all the MR and adaptor types in RMM for different kinds of memory.

The problem with @tfeher 's suggestion at the top is that it appears to mix host- and device-accessible memory kinds under `device_memory_resource`, which is supposed to allocate only device-accessible memory.

As to @cjnolet 's comment, I agree that you should have separate resources for different kinds of allocation, and then have a higher level MR that takes pointers/refs to MRs for each kind you want to dispatch to, rather than hard-coding different allocation calls within a single MR.

**Review by tfeher — APPROVED** [review]

**Pull request merged**
